plugin/trino-kudu/src/test/java/io/trino/plugin/kudu/TestingKuduServer.java

```diff
@@ -15,65 +15,184 @@
+import static org.testcontainers.utility.MountableFile.forClasspathResource;
 
 public class TestingKuduServer
         implements Closeable
 {
+    private String hostTmpDirectory;
+    private String masterTmpDirectory;
+    private List<String> tabletTmpDirectories;
+
     private static final Integer KUDU_MASTER_PORT = 7051;
     private static final Integer KUDU_TSERVER_PORT = 7050;
     private static final Integer NUMBER_OF_REPLICA = 3;
 
     private final GenericContainer<?> master;
     private final List<GenericContainer<?>> tServers;
 
+    /**
+     * WORKAROUND WARNING:
+     * Kudu needs to know the host port it will be bound to in order to configure --rpc_advertised_addresses
+     * However when using non-fixed ports in testcontainers, we only know the host/mapped port after the container starts up
+     * In order to workaround this, we create volume mounts into each container and startup each container and have them watch the volume mounts
+     * When the container starts up, we get the mapped port and write into the volume mount a special "mapped_port" file saying what the mapped port is
+     * Then we can configure --rpc_advertised_addresses and startup kudu
+     * Finally, since we never actually checked the containers started up correctly, we manually fire the HostPortWaitStrategy
+     */
     public TestingKuduServer()
     {
         Network network = Network.newNetwork();
         ImmutableList.Builder<GenericContainer<?>> tServersBuilder = ImmutableList.builder();
-        this.master = new GenericContainer<>("apache/kudu:1.10.0")
+
+        String hostIP = getHostIPAddress();
+        createHostTmpDirectoriesForVolumeMounts();
+
+        String masterContainerAlias = "kudu-master";
+        this.master = new GenericContainer<>("apache/kudu:1.15.0")
+                .withFileSystemBind(masterTmpDirectory, "/testcontainers")
+                .withCopyFileToContainer(forClasspathResource("wait_until_port_bound.sh", 0777), "/wait_until_port_bound.sh")
                 .withExposedPorts(KUDU_MASTER_PORT)
-                .withCommand("master")
+                .withCommand("/wait_until_port_bound.sh", "master")
+                .withEnv("KUDU_MASTERS", "kudu-master:7051")
+                // Note `rpc_advertised_addresses` does not have a port configured, it will be configured after container startup in `/wait_until_port_bound.sh`
+                .withEnv("MASTER_TEMPLATE_ARGS", format("--fs_wal_dir=/var/lib/kudu/master --logtostderr --use_hybrid_clock=false --rpc_bind_addresses=%s:%s --rpc_advertised_addresses=%s:", masterContainerAlias, KUDU_MASTER_PORT, hostIP))
                 .withNetwork(network)
-                .withNetworkAliases("kudu-master");
+                .withNetworkAliases(masterContainerAlias)
+                // Do not wait, the host port will not be listening until after the container starts up and gets the mapped host port
+                .waitingFor(new NoWaitStrategy());
```

Comment:
Do we need this fix for Master server too ? I think we don't need `rpc_advertised_addresses` for master server

Reply:
Yes you are correct 👍